### ISSUE TITLE
magicgui.widgets.CategoricalWidget not found in magicgui 0.2.1

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/magicgui#81]
**Describe the bug**
Hi Talley @tlambert03,

tl;dr: Is there example code available for using choices in magicgui 0.2.1? I get error messages while attempting to use `CategoricalWidget`.

I'm just updating the magicgui dependency in [my experimental napari plugin](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_operations/_operations.py#L12) and receive this error message in the context of having a `choice`:

```
As of magicgui 0.2.0, a `choices` callable may accept only a single positional
argument (an instance of `magicgui.widgets.CategoricalWidget`), and must return
an iterable (the choices to show). Function 'get_layers' accepts 2 arguments.
In the future, this will raise an exception.
```

However, the mentioned class `magicgui.widgets.CategoricalWidget` does not exist. If I import it, I receive this error message:

```
(clustering) C:\structure\code\napari_pyclesperanto_assistant\docs\demo>python start_assistant.py
C:\Users\rober\miniconda3\envs\clustering\lib\site-packages\magicgui\_qt\widgets.py:13: FutureWarning:

The 'magicgui._qt.widgets' module has been removed.
For 'QDataComboBox', please use 'magicgui.widgets.ComboBox' directly.
(Or just use the string `widget_type='ComboBox'`)
In the future this will raise an exception.

  warnings.warn(
I'm asked:  C:\structure\code\napari_pyclesperanto_assistant\napari_pyclesperanto_assistant\data\Lund_000500_resampled-cropped.tif
ERROR:root:Unhandled exception:
Traceback (most recent call last):
  File "C:\Users\rober\miniconda3\envs\clustering\lib\site-packages\napari\_qt\event_loop.py", line 79, in gui_qt
    yield app
  File "start_assistant.py", line 12, in <module>
    assistant_gui = napari_pyclesperanto_assistant.napari_plugin(viewer)
  File "c:\structure\code\napari_pyclesperanto_assistant\napari_pyclesperanto_assistant\_gui\_napari_plugin.py", line 6, in napari_plugin
    gui = AssistantGUI(viewer)
  File "c:\structure\code\napari_pyclesperanto_assistant\napari_pyclesperanto_assistant\_gui\_AssistantGui.py", line 26, in __init__
    self._init_gui()
  File "c:\structure\code\napari_pyclesperanto_assistant\napari_pyclesperanto_assistant\_gui\_AssistantGui.py", line 37, in _init_gui
    from .._operations._operations import denoise, background_removal, filter, binarize, combine, label, label_processing, map, mesh, measure
  File "c:\structure\code\napari_pyclesperanto_assistant\napari_pyclesperanto_assistant\_operations\_operations.py", line 6, in <module>
    from magicgui.widgets import CategoricalWidget
ImportError: cannot import name 'CategoricalWidget' from 'magicgui.widgets' (C:\Users\rober\miniconda3\envs\clustering\lib\site-packages\magicgui\widgets\__init__.py)
```

**To Reproduce**

```
pip install magicgui==0.2.1

git clone https://github.com/clEsperanto/napari_pyclesperanto_assistant

cd napari_pyclesperanto_assistant/docs/demo

python start_assistant.py
```

for the second error message, add `from magicgui.widgets import CategoricalWidget` to [this file](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_operations/_operations.py)

**Expected behavior**

**Screenshots**

**Environment (please complete the following information):**
 - Windows
 - magicgui version 0.2.1

```
(clustering) C:\structure\code\napari_pyclesperanto_assistant\docs\demo>conda list
# packages in environment at C:\Users\rober\miniconda3\envs\clustering:
#
# Name                    Version                   Build  Channel
alabaster                 0.7.12                   pypi_0    pypi
appdirs                   1.4.4                    pypi_0    pypi
babel                     2.9.0                    pypi_0    pypi
backcall                  0.2.0                    pypi_0    pypi
blas                      1.0                         mkl
ca-certificates           2020.12.8            haa95532_0
cachey                    0.2.1                    pypi_0    pypi
certifi                   2020.12.5        py38haa95532_0
chardet                   4.0.0                    pypi_0    pypi
colorama                  0.4.4                    pypi_0    pypi
cudatoolkit               11.0.221             h74a9793_0
cycler                    0.10.0                   pypi_0    pypi
dask                      2020.12.0                pypi_0    pypi
decorator                 4.4.2                    pypi_0    pypi
docutils                  0.16                     pypi_0    pypi
freetype-py               2.2.0                    pypi_0    pypi
heapdict                  1.0.1                    pypi_0    pypi
idna                      2.10                     pypi_0    pypi
imageio                   2.9.0                    pypi_0    pypi
imagesize                 1.2.0                    pypi_0    pypi
intel-openmp              2020.2                      254
ipykernel                 5.4.2                    pypi_0    pypi
ipython                   7.19.0                   pypi_0    pypi
ipython-genutils          0.2.0                    pypi_0    pypi
jedi                      0.18.0                   pypi_0    pypi
jinja2                    2.11.2                   pypi_0    pypi
joblib                    1.0.0                    pypi_0    pypi
jupyter-client            6.1.7                    pypi_0    pypi
jupyter-core              4.7.0                    pypi_0    pypi
kiwisolver                1.3.1                    pypi_0    pypi
libuv                     1.40.0               he774522_0
magicgui                  0.2.1                    pypi_0    pypi
markupsafe                1.1.1                    pypi_0    pypi
matplotlib                3.3.3                    pypi_0    pypi
mkl                       2020.2                      256
mkl-service               2.3.0            py38h196d8e1_0
mkl_fft                   1.2.0            py38h45dec08_0
mkl_random                1.1.1            py38h47e9c7a_0
napari                    0.4.2                    pypi_0    pypi
napari-plugin-engine      0.1.8                    pypi_0    pypi
napari-pyclesperanto-assistant 0.2.1                     dev_0    <develop>
napari-svg                0.1.4                    pypi_0    pypi
networkx                  2.5                      pypi_0    pypi
ninja                     1.10.2           py38h6d14046_0
numpy                     1.19.3                   pypi_0    pypi
numpydoc                  1.1.0                    pypi_0    pypi
openssl                   1.1.1i               h2bbff1b_0
packaging                 20.8                     pypi_0    pypi
parso                     0.8.1                    pypi_0    pypi
pickleshare               0.7.5                    pypi_0    pypi
pillow                    8.0.1                    pypi_0    pypi
pip                       20.3.1           py38haa95532_0
prompt-toolkit            3.0.8                    pypi_0    pypi
psutil                    5.8.0                    pypi_0    pypi
pyclesperanto-prototype   0.6.0                     dev_0    <develop>
pygments                  2.7.3                    pypi_0    pypi
pyopencl                  2020.3.1+cl12            pypi_0    pypi
pyopengl                  3.1.5                    pypi_0    pypi
pyparsing                 2.4.7                    pypi_0    pypi
pyperclip                 1.8.1                    pypi_0    pypi
pyqt5                     5.15.2                   pypi_0    pypi
pyqt5-sip                 12.8.1                   pypi_0    pypi
python                    3.8.5                h5fd99cc_1
python-dateutil           2.8.1                    pypi_0    pypi
pytools                   2020.4.4                 pypi_0    pypi
pytorch                   1.7.1           py3.8_cuda110_cudnn8_0    pytorch
pytz                      2020.5                   pypi_0    pypi
pywavelets                1.1.1                    pypi_0    pypi
pywin32                   300                      pypi_0    pypi
pyyaml                    5.3.1                    pypi_0    pypi
pyzmq                     20.0.0                   pypi_0    pypi
qtconsole                 5.0.1                    pypi_0    pypi
qtpy                      1.9.0                    pypi_0    pypi
requests                  2.25.1                   pypi_0    pypi
scikit-image              0.18.0                   pypi_0    pypi
scikit-learn              0.24.0                   pypi_0    pypi
scipy                     1.5.4                    pypi_0    pypi
setuptools                51.0.0           py38haa95532_2
six                       1.15.0           py38haa95532_0
snowballstemmer           2.0.0                    pypi_0    pypi
sphinx                    3.4.1                    pypi_0    pypi
sphinxcontrib-applehelp   1.0.2                    pypi_0    pypi
sphinxcontrib-devhelp     1.0.2                    pypi_0    pypi
sphinxcontrib-htmlhelp    1.0.3                    pypi_0    pypi
sphinxcontrib-jsmath      1.0.1                    pypi_0    pypi
sphinxcontrib-qthelp      1.0.3                    pypi_0    pypi
sphinxcontrib-serializinghtml 1.1.4                    pypi_0    pypi
sqlite                    3.33.0               h2a8f88b_0
threadpoolctl             2.1.0                    pypi_0    pypi
tifffile                  2020.12.8                pypi_0    pypi
toolz                     0.11.1                   pypi_0    pypi
tornado                   6.1                      pypi_0    pypi
traitlets                 5.0.5                    pypi_0    pypi
typing_extensions         3.7.4.3                    py_0
urllib3                   1.26.2                   pypi_0    pypi
vc                        14.2                 h21ff451_1
vispy                     0.6.6                    pypi_0    pypi
vs2015_runtime            14.27.29016          h5e58377_2
wcwidth                   0.2.5                    pypi_0    pypi
wheel                     0.36.1             pyhd3eb1b0_0
wincertstore              0.2                      py38_0
wrapt                     1.12.1                   pypi_0    pypi
zlib                      1.2.11               h62dcd97_4
```

**Comment by tlambert03** [comment]
Hey @haesleinhuepf, thanks for the detailed report and sorry for the trouble. Some quick answers here, but let me also dig into the assistant repo a bit more, submit some PRs, and maybe we can hop on a zoom call to make sure everything is going ok with the new releases.

tl;dr: most of your issues here are on the napari side.  If I do pip `install git+https://github.com/napari/napari.git`, it mostly starts working again.  And that error message was directed at napari, not at you.

> tl;dr: Is there example code available for using choices in magicgui 0.2.1? I get error messages while attempting to use CategoricalWidget.  I'm just updating the magicgui dependency in my experimental napari plugin and receive this error message in the context of having a choice:

Probably the biggest issue here is that warning message is directed at napari (the user of the `register_type` function), and _not_ at you the user ... but that's not quite clear enough.  The hint is in ` Function 'get_layers' accepts 2 arguments.`... and `get_layers` is not your function.  But I can make that warning message clearer and add additional info so that users of a third-party magicgui provider know it's not their issue.  The point is, napari (0.4.2) needed to update [this `get_layers` function](https://github.com/napari/napari/blob/7ba2d48db285c762096cc5099ebf5278af36762d/napari/utils/_magicgui.py#L78) ... which was done in https://github.com/napari/napari/pull/1994, so once you update napari, that warning should go away.


> tl;dr: Is there example code available for using choices in magicgui 0.2.1? I get error messages while attempting to use CategoricalWidget. 

The API for choices hasn't changed.  Your [usage here](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_operations/_operations.py#L12) (i.e. a list of strings) is just fine and will continue to work.  The warning message there for the user of `magicgui.register_type` (napari).  I'll improve the warning message, and I can add a dedicated example for using choices.  As with <0.2.0, there are three valid arguments to `choices`: a list of strings, an enum, or a callable that returns an iterable of valid choices.

> However, the mentioned class `magicgui.widgets.CategoricalWidget` does not exist. If I import it, I receive this error message:

sorry again, will update the warning message.  That's a base widget that lives at `magicgui.widgets._bases.CategoricalWidget`... It's not meant to be used directly, but it's telling the developer that a choices _callable_ (which, again, napari provided, not you... you provided a list of strings) will receive an _instance_ of a `CategoricalWidget` subclass.  Not that you need to provide one directly.

Long story short... I'm worried you're getting the impression that a ton of stuff has changed on the user end, and it really hasn't.  There was just some two-repo version juggling that should be mostly resolved now with magicgui >= 0.2.1 and napari https://github.com/napari/napari/pull/1994/ and https://github.com/napari/napari/pull/1856

as mentioned, I'll fork your repo and start submitting PRs to help out. but in the meantime, try updating napari, and then we can narrow down the few things that you'll need to change on your end

**Comment by haesleinhuepf** [comment]
Thanks for the detailed explanation and for your time @tlambert03 ! I know how it feels to make these breaking changes. I also think your timing is good: Better now than in half a year when there are a hundred napari-plugins floating around which all make use of magicgui ;-)

I'd be happy to have a zoom with you at some point, also because I think I used magicgui in a bit unusual / excessive way. I could imagine that some things I implemented might be better hosted on magicgui-side. I'm also aware that my use-case might not be a general use case though. You've already seen/answered my comments [here](https://github.com/napari/napari/pull/1816#issuecomment-754712879).

**Comment by tlambert03** [comment]
yep.  It had to be done.
I'm going through your repo now.  It doesn't take much to make it compatible with 0.2.1 and napari 0.4.3... and I think all the changes feel "right" too.  will PR soon.

**Comment by tlambert03** [comment]
I also added an example of all the valid ways to supply options to a dropdown menu:
https://github.com/napari/magicgui/blob/master/examples/choices.py

**Comment by tlambert03** [comment]
If it's ok with you, I'm going to close this issue.  I clarified the warning message in #83, added more examples in #85.  Thanks again for the very helpful feedback!  keep it coming.
